

\texttt{/lib/accuracy.pl} \\
%
\begin{itemize}
\item accuracy\_stats(++RefFunctor,++PredFunctor,++Start,++End,--OutputFile)
\item sensitivity(++TruePos,++FalseNeg,--Sens)
\item specificity(++TruePos,++FalsePos,--Spec)
\item specificity_traditional(++TrueNeg,++FalsePos,--ST)
\item correlation_coefficient(++TruePos,++FalsePos,++TrueNeg,++FalseNeg,--CC)
\item simple_matching_coefficient(++TruePos,++FalsePos,++TrueNeg,++FalseNeg,--SMC)
\item average_conditional_probability(++TruePos,++FalsePos,++TrueNeg,++FalseNeg,--ACP)
\item approximate_correlation(++TruePos,++FalsePos,++TrueNeg,++FalseNeg,--AC)
\item count\_genes(++Functor,++Start,++End,--Count)
\item correct\_genes(++RefFunctor,++PredFunctor,++Start,++End,--Correct\_Genes)
\item missing\_genes(++RefFunctor,++PredFunctor,++Start,++End,--Missing\_Genes)
\item wrong\_genes(++RefFunctor,++PredFunctor,++Start,++End,--Wrong\_Genes)
\item number\_of\_correct\_genes(++RefFunctor,++PredFunctor,++Start,++End,--Num\_Correct)
\item number\_of\_missing\_genes(++RefFunctor,++PredFunctor,++Start,++End,--Num\_Missing)
\item number\_of\_wrong\_genes(++RefFunctor,++PredFunctor,++Start,++End,--Num\_Wrong)
\item gene\_finding\_difficulty\_report(++RefFunctor,++PreducFunctors,++RangeMin,++RangeMax,++MatchCriteria,--GenePredictionDifficultyList)
\item gene\_difficulty\_score(++NumGeneFinders,++NumFoundGene,--DifficultyScore)
\item report\_nstats(++RefFunctor,++PredicFunctor,++Start,++End)
\item strand(??Strand)
\item reading\_frame(??Frame)
\end{itemize}

\texttt{/lib/autoAnnotations.pl} \\
%
\begin{itemize}
\item prismAnnot(++File)
\item prismAnnot(++File,++CompilationOptions)
\item viterbiAnnot(++Goal,--Proba)
\end{itemize}


\texttt{/lib/chmm.pl} \\
%
\begin{itemize}
\item init\_store
\item init\_constraint\_store(++Constraint,--Store)
\item forward\_store(++Store)
\item get\_store(++S)
\item init_constraint_stores(++Constraints,--Stores)
\item init_constraint_store(++Constraint,--Store)
\item check\_constraints(++StateUpdate)
\item check\_each\_constraints(++StateUpdate,++Constraints,++StoreBefores,--StoreAfters)
\item constraint\_check(++Constraint,++StateUpdate,++StoreBefore,--StoreAfter)
\end{itemize}


\texttt{/lib/fasta.pl}\\
%
\begin{itemize}
\item Maybe to merge with the fna parser ?
\item fasta\_load\_sequence(++InputFile,++SequenceId,--FastaHeader,--Sequence) 
\item fasta\_save\_sequence(++OutputFile,++SequenceData,++Header) 
\item atom\_code\_list(++Codes,,--Atoms)
\end{itemize}

\texttt{/lib/genecode.pl} \\
%
\begin{itemize}
\item genecode(++TableNum,??Codon,??Amino\_Acid)
\item genecode\_start\_codon(++TableNum,??Codon,??Amino\_Acid)
\item genecode\_stop\_codon(++TableNum,??Codon)
\item genecode\_start\_codons(++GeneCode,--StartCodons)
\item genecode\_stop\_codons(++GeneCode,--StopCodons)
\end{itemize}

\texttt{/lib/genedb.pl} \\
%
\begin{itemize}
\item To remove ??
\end{itemize}

\texttt{/lib/interface.pl} \\
%
\begin{itemize}
\item >(++Call)
\item get\_annotation\_file(++Model,++Inputs,++Options,--FileName)
\item run\_model(++Model,++Inputs,++Options,--FileName)
\item train\_model(++Model,++TrainingDataFiles,++Options,--SavedParamsFile)
\item get\_option(++Options,--KeyValue)
\item get\_option(++Options,--Key, --Value)
\item launch\_prism\_process(++PrismFile,--Goal)
\item lost\_tmp\_directory(--Dir)
\item lost\_models\_directory(--Dir)
\item lost\_data\_directory(--Dir)
\item lost\_model\_directory(++Model,--Dir)
\item lost\_model\_parameters\_directory(++Model,--Dir)
\item lost\_model\_annotations\_directory(++Model,--Dir) % Used ?
\item lost\_model\_interface\_file(++Model,--ModelFile)
\item lost\_model\_parameter\_index\_file(++Model,--IndexFile)
\item lost\_annotation\_index\_file(++Index)
\item lost\_data\_file(++SequenceID,--SequenceFile)
\item is\_generated\_file(++File)
\item lost\_file\_index\_get\_filename(++Index,++Model,++InputFiles,++Options,--Filename)
\item check\_valid\_model\_call(++Model,++InterfacePredicate,++Arity,++Options)
\item declared\_model\_options(++Model,++Goal,--DeclaredOptions)
\item lost\_interface\_input\_formats(++Model,++InterfacePredicate,--Formats)
\item lost\_interface\_defines\_output\_format(++Model,--InterfacePredicate)
\item lost\_model\_option\_values(++Model,++InterfacePredicate,--OptionName,--Value)  
\item rm\_gen
\item rm\_tmp
\item list\_lost\_models(--Models) 
\end{itemize}




\texttt{/lib/misc\_utils.pl} \\
%
\begin{itemize}
\item atom\_concat\_list(++List,--Atom)
\item atom\_list\_code\_list(++Atoms,--Codes)
\item inlists\_nth0(++List,++Nth,--ListElt)
\item flexible\_append(++List1,++List2,--List)
\item flatten\_once(++List,--ListFlatten)
\item map(++Functor,++InList,--OutList) :-
\item list\_head(++List,--Head)
\item list\_tail(++List,--Tail)
\item rotate\_list\_vector(++ListOfLists,--RotatedList) 
\item replace(++Symbol,++Replacement,++Inlist,--Outlist)
\item match\_tail(++InputList,--HeadOfInputList,++TailOfInputList)
\item not\_member(++Elt,++List)
\item terms\_has\_rule\_with\_head(++Terms,++Functor,Arity)
\item abs(++Elt,--Abs)  % Not ISO predicate ?        
\item min(++Elt1,++El2,--Min)   % Not ISO predicate ?
\item max(++Elt1,++Elt2,--Max)  % Not ISO predicate ?
\item list\_max(++List,--Max)
\item list\_min(++List,--Min)
\item atom2integer(++Atom,--Integer)
\item check\_or\_fail(++Goal,++Error)
\item check\_or\_warn(++Check,++Error)
\item dirname(++Filename,--DirPartAtom)
\item overlaps(++Start1,++End1,++Start2,++End2)
\item sum\_range\_list(++Ranges,--Sum)
\item upper\_lower(??UpperLower,??UpperLower)
\item is\_upper\_case\_alphanumeric(??Code)
\item is\_lower\_case\_alphanumeric(??Code)
\end{itemize}
